### PR TITLE
Small Bugfix: Remove possibility of bitshifting negative integers

### PR DESCRIPTION
**Pull request by lroberts36**
## PR Summary

This PR fixes the bitshift related undefined behavior described in #1107 by multiplying by a power of two explicitly rather than performing the same operation (on positive numbers) via a bitshift.

## PR Checklist

- [ ] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [ ] Code is formatted
- [ ] Changes are summarized in CHANGELOG.md
- [ ] Change is breaking (API, behavior, ...)
  - [ ] Change is *additionally* added to CHANGELOG.md in the breaking section
  - [ ] PR is marked as breaking
  - [ ] Short summary API changes at the top of the PR (plus optionally with an automated update/fix script)
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [ ] Docs build
- [ ] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

File: src/mesh/forest/logical_location.cpp [hunk 1]
```diff
@@ -118,10 +118,10 @@ bool LogicalLocation::IsNeighbor(const LogicalLocation &in) const {
 
   bool neighbors = true;
   for (int dir = 0; dir < 3; ++dir) {
-    auto low = (l(dir) << level_shift_this) - 1;
+    auto low = l(dir) * block_size_this - 1;
     auto hi = low + block_size_this + 1;
 
-    auto low_in = (in.l(dir) << level_shift_in);
+    auto low_in = in.l(dir) * block_size_in;
     auto hi_in = low_in + block_size_in - 1;
```
Comment (pgrete) [inline src/mesh/forest/logical_location.cpp:125]: Out of curiosity, where does the asymmetry come from (i.e., that low/hi is offset by-1/+1 and low_in, hi_in by 0,-1)?
Would be great to also add a short comment in the code.

**Review by pgrete — APPROVED** [review]
Why didn't case this bigger issues? Was this an edge case or working by chance?

**Comment by pgrete** [comment]
btw you got the Schnapszahl ;)

**Comment by lroberts36** [comment]
@pgrete: I am not sure why this worked previously, but I did not look carefully at what happened with the negative but shift. I think this should have been hit reasonably often.

Haha, I had to look up schnapszahl...

**Review by BenWibking — APPROVED** [review]
The new code looks fine to me, but I am likewise confused as to how the cases where it was bitshifting -1 worked...

**Review by Yurlungur — APPROVED** [review]

File: src/mesh/forest/logical_location.cpp (hunk 1 above)
Comment (lroberts36) [reply src/mesh/forest/logical_location.cpp:125]: It just has to do with how the overlap is checked, i.e. the bounds of `this` are increased by one in the positive and negative direction to create a one layer thick ghost region.

**Comment by pgrete** [comment]
The test now reliably failed twice. Does anyone see how this changeset could result in that error? I don't.

**Comment by BenWibking** [comment]
> The test now reliably failed twice. Does anyone see how this changeset could result in that error? I don't.

It's a NumPy 2.0 API breakage 😞...
```
AttributeError: module 'numpy' has no attribute 'product'
```

**Comment by BenWibking** [comment]
The test scripts need to be updated: https://numpy.org/devdocs/numpy_2_0_migration_guide.html#numpy-2-migration-guide

**Comment by lroberts36** [comment]
@pgrete and @BenWibking: Apparently this worked because (at least clang and gcc) bitshift negative numbers in the expected way, i.e `neg_num << n == neg_num * std::pow(2, n)`.

**Comment by BenWibking** [comment]
CI fix here: https://github.com/parthenon-hpc-lab/parthenon/pull/1116

**Comment by pgrete** [comment]
> @pgrete and @BenWibking: Apparently this worked because (at least clang and gcc) bitshift negative numbers in the expected way, i.e `neg_num << n == neg_num * std::pow(2, n)`.

great, thanks for confirming!

**Pull request merged**